Add support for expanding relevant VS Code Predefined Variables

bytehow opened the pull request:
- Updated launch.json to support multifolder and single folder configurations.
- Added multi-folder workspace configuration file with terminal link matchers.
- Added example regexes to test against.
- Made link tooltip rendering show the _URL dedcoded_ URL

This feature is very useful when trying to commit shared rules into a repository, without hard coding a full path.

Fixes #1

bytehow committed:
Add support for expanding relevant VS Code Predefined Variables
- Updated launch.json to support multifolder and single folder configurations.
- Added multi-folder workspace configuration file with terminal link matchers.
- Added example regexes to test against.

bytehow commented:
Some example output.

## Single folder:
```
2025-07-10 09:41:58.086 [info] Activated
2025-07-10 09:41:58.086 [info] Parsing config:
[
  {
    "regex": "/(home|opt|Users)\\/[^\\s:]+\\/[^\\s:]+\\.(py|c|rs|cpp)",
    "uri": "file:${workspaceFolder}/$0"
  },
  {
    "regex": "test_userHome",
    "uri": "file:${userHome}/test.txt"
  },
  {
    "regex": "test_workspaceFolderBasename",
    "uri": "file:${workspaceFolderBasename}/test.txt"
  },
  {
    "regex": "test_cwd",
    "uri": "file:cwd:'${cwd}'/test.txt"
  },
  {
    "regex": "test_pathSeparator",
    "uri": "file:sep${pathSeparator}"
  },
  {
    "regex": "test_slash",
    "uri": "file:slash${/}"
  },
  {
    "regex": "test_envVAR",
    "uri": "file:${env:HOME}/test.txt"
  },
  {
    "regex": "test_workspaceFolder_Single",
    "uri": "file:${workspaceFolder:single-folder}/test.txt"
  },
  {
    "regex": "test_workspaceFolderBasename_Single",
    "uri": "file:${workspaceFolderBasename:single-folder}/test.txt"
  }
]
2025-07-10 09:41:58.086 [trace] [expandVariables] input: file:${workspaceFolder}/$0
2025-07-10 09:41:58.086 [trace] [expandVariables] result: file:/Users/bytehow/code/terminal-links/test/single-folder/$0
2025-07-10 09:41:58.086 [trace] [expandVariables] input: file:${userHome}/test.txt
2025-07-10 09:41:58.086 [trace] [expandVariables] result: file:/Users/bytehow/test.txt
2025-07-10 09:41:58.086 [trace] [expandVariables] input: file:${workspaceFolderBasename}/test.txt
2025-07-10 09:41:58.086 [trace] [expandVariables] result: file:single-folder/test.txt
2025-07-10 09:41:58.086 [trace] [expandVariables] input: file:cwd:'${cwd}'/test.txt
2025-07-10 09:41:58.086 [trace] [expandVariables] result: file:cwd:'/'/test.txt
2025-07-10 09:41:58.086 [trace] [expandVariables] input: file:sep${pathSeparator}
2025-07-10 09:41:58.086 [trace] [expandVariables] result: file:sep/
2025-07-10 09:41:58.086 [trace] [expandVariables] input: file:slash${/}
2025-07-10 09:41:58.086 [trace] [expandVariables] result: file:slash/
2025-07-10 09:41:58.086 [trace] [expandVariables] input: file:${env:HOME}/test.txt
2025-07-10 09:41:58.086 [trace] [expandVariables] result: file:/Users/bytehow/test.txt
2025-07-10 09:41:58.086 [trace] [expandVariables] input: file:${workspaceFolder:single-folder}/test.txt
2025-07-10 09:41:58.086 [trace] [expandVariables] result: file:/Users/bytehow/code/terminal-links/test/single-folder/test.txt
2025-07-10 09:41:58.086 [trace] [expandVariables] input: file:${workspaceFolderBasename:single-folder}/test.txt
2025-07-10 09:41:58.086 [trace] [expandVariables] result: file:single-folder/test.txt
2025-07-10 09:41:58.086 [trace] Parsed config:
  {
    regex: "\/(home|opt|Users)\/[^\s:]+\/[^\s:]+\.(py|c|rs|cpp)",
    uri: "file:/Users/bytehow/code/terminal-links/test/single-folder/$0"
  }
  {
    regex: "test_userHome",
    uri: "file:/Users/bytehow/test.txt"
  }
  {
    regex: "test_workspaceFolderBasename",
    uri: "file:single-folder/test.txt"
  }
  {
    regex: "test_cwd",
    uri: "file:cwd:'/'/test.txt"
  }
  {
    regex: "test_pathSeparator",
    uri: "file:sep/"
  }
  {
    regex: "test_slash",
    uri: "file:slash/"
  }
  {
    regex: "test_envVAR",
    uri: "file:/Users/bytehow/test.txt"
  }
  {
    regex: "test_workspaceFolder_Single",
    uri: "file:/Users/bytehow/code/terminal-links/test/single-folder/test.txt"
  }
  {
    regex: "test_workspaceFolderBasename_Single",
    uri: "file:single-folder/test.txt"
  }
  ```
  
  ## Multi folder
  
  ```
  2025-07-10 09:43:10.719 [info] Activated
2025-07-10 09:43:10.719 [info] Parsing config:
[
  {
    "regex": "/(home|opt|Users)\\/[^\\s:]+\\/[^\\s:]+\\.(py|c|rs|cpp)",
    "uri": "file:${workspaceFolder}/$0"
  },
  {
    "regex": "test_userHome",
    "uri": "file:${userHome}/test.txt"
  },
  {
    "regex": "test_workspaceFolderBasename",
    "uri": "file:${workspaceFolderBasename}/test.txt"
  },
  {
    "regex": "test_cwd",
    "uri": "file:cwd:'${cwd}'/test.txt"
  },
  {
    "regex": "test_pathSeparator",
    "uri": "file:sep${pathSeparator}"
  },
  {
    "regex": "test_slash",
    "uri": "file:slash${/}"
  },
  {
    "regex": "test_envVAR",
    "uri": "file:${env:HOME}/test.txt"
  },
  {
    "regex": "test_workspaceFolder_Main",
    "uri": "file:${workspaceFolder:Main Folder}/test.txt"
  },
  {
    "regex": "test_workspaceFolder_Another",
    "uri": "file:${workspaceFolder:Another Folder}/test.txt"
  },
  {
    "regex": "test_workspaceFolderBasename_Main",
    "uri": "file:${workspaceFolderBasename:Main Folder}/test.txt"
  },
  {
    "regex": "test_workspaceFolderBasename_Another",
    "uri": "file:${workspaceFolderBasename:Another Folder}/test.txt"
  }
]
2025-07-10 09:43:10.719 [trace] [expandVariables] input: file:${workspaceFolder}/$0
2025-07-10 09:43:10.719 [trace] [expandVariables] result: file:/Users/bytehow/code/terminal-links/test/multi-folder/main-folder/$0
2025-07-10 09:43:10.719 [trace] [expandVariables] input: file:${userHome}/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] result: file:/Users/bytehow/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] input: file:${workspaceFolderBasename}/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] result: file:main-folder/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] input: file:cwd:'${cwd}'/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] result: file:cwd:'/'/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] input: file:sep${pathSeparator}
2025-07-10 09:43:10.719 [trace] [expandVariables] result: file:sep/
2025-07-10 09:43:10.719 [trace] [expandVariables] input: file:slash${/}
2025-07-10 09:43:10.719 [trace] [expandVariables] result: file:slash/
2025-07-10 09:43:10.719 [trace] [expandVariables] input: file:${env:HOME}/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] result: file:/Users/bytehow/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] input: file:${workspaceFolder:Main Folder}/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] result: file:/Users/bytehow/code/terminal-links/test/multi-folder/main-folder/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] input: file:${workspaceFolder:Another Folder}/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] result: file:/Users/bytehow/code/terminal-links/test/multi-folder/another-folder/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] input: file:${workspaceFolderBasename:Main Folder}/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] result: file:main-folder/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] input: file:${workspaceFolderBasename:Another Folder}/test.txt
2025-07-10 09:43:10.719 [trace] [expandVariables] result: file:another-folder/test.txt
2025-07-10 09:43:10.719 [trace] Parsed config:
  {
    regex: "\/(home|opt|Users)\/[^\s:]+\/[^\s:]+\.(py|c|rs|cpp)",
    uri: "file:/Users/bytehow/code/terminal-links/test/multi-folder/main-folder/$0"
  }
  {
    regex: "test_userHome",
    uri: "file:/Users/bytehow/test.txt"
  }
  {
    regex: "test_workspaceFolderBasename",
    uri: "file:main-folder/test.txt"
  }
  {
    regex: "test_cwd",
    uri: "file:cwd:'/'/test.txt"
  }
  {
    regex: "test_pathSeparator",
    uri: "file:sep/"
  }
  {
    regex: "test_slash",
    uri: "file:slash/"
  }
  {
    regex: "test_envVAR",
    uri: "file:/Users/bytehow/test.txt"
  }
  {
    regex: "test_workspaceFolder_Main",
    uri: "file:/Users/bytehow/code/terminal-links/test/multi-folder/main-folder/test.txt"
  }
  {
    regex: "test_workspaceFolder_Another",
    uri: "file:/Users/bytehow/code/terminal-links/test/multi-folder/another-folder/test.txt"
  }
  {
    regex: "test_workspaceFolderBasename_Main",
    uri: "file:main-folder/test.txt"
  }
  {
    regex: "test_workspaceFolderBasename_Another",
    uri: "file:another-folder/test.txt"
  }

  ```

bytehow committed:
Fix URI encoding in tooltips for terminal links

Mossop commented on src/extension.ts:
```diff
@@ -152,3 +156,45 @@ export function activate(context: ExtensionContext) {
+  const wsPath = wsFolder?.uri.fsPath || "";
+  const wsBasename = wsFolder ? path.basename(wsFolder.uri.fsPath) : "";
+  const userHome = process.env.HOME || process.env.USERPROFILE || "";
+  const cwd = process.cwd();
```
What you've done here just gives us the working directory for the VSCode process, on macOS this is just `/`. I had thought we could use the working directory of the terminal, which is [available](https://code.visualstudio.com/api/references/vscode-api#1959) but that would be more work because we would have to do the expansion when matching rather than at this config parsing phase. It also changes dynamically which the terminal matching API doesn't really support.

So I think we just shouldn't add support for `cwd`, at least not right now.

bytehow replied:
Good cathc! Totally agree. Removed. I thought I had invoked vscode wrong somehow when I saw that `cwd` was evaluting to `/`

Mossop reviewed (changes requested):
This looks great, thanks. One thing I think we should remove and if you wouldn't mind fixing the lint errors that would be good, otherwise I'll just take care of those at some point.

bytehow committed:
Fix linter errors

bytehow committed:
Remove ${cwd} variable expansion

bytehow commented:
Sorry for the late response. I didn't realize my GitHub email notifications were misconfigured, so I didn't see your comment! I've fixed the lint errors, and removed `${cwd}`

Mossop reviewed (approved):
Looks good, thanks
